lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs

```diff
@@ -0,0 +1,944 @@
+unMockWid :: MWid -> WalletId
+unMockWid (MWid wid) = WalletId . hash . B8.pack $ wid
+
+widPK' :: Expr WalletId -> PrimaryKey WalletId
+widPK' = PrimaryKey . evalWalletId
+
+-- | Represent (XPrv, Hash) as a string.
+type MPrivKey = String
+
+-- | Stuff a mock private key into the type used by DBLayer.
+fromMockPrivKey :: MPrivKey -> (Key purpose XPrv, Hash "encryption")
+fromMockPrivKey s = (k, Hash (B8.pack s))
+    where Right (k, _) = deserializeXPrv (B8.replicate 256 '0', mempty)
+
+-- | Unstuff the DBLayer private key into the mock type.
+toMockPrivKey :: (Key purpose XPrv, Hash "encryption") -> MPrivKey
+toMockPrivKey (_, Hash h) = B8.unpack h
+
+data Mock = M
+    { checkpoints :: Map MWid MWallet
+    , metas :: Map MWid WalletMetadata
+    , txs :: Map MWid TxHistory
+    , privateKey :: Map MWid MPrivKey
+    } deriving (Show, Generic)
+
+emptyMock :: Mock
+emptyMock = M mempty mempty mempty mempty
+
+type MockOp a = Mock -> (Either (Err MWid) a, Mock)
+
+mCleanDB :: MockOp ()
+mCleanDB _ = (Right (), emptyMock)
+
+mCreateWallet :: MWid -> MWallet -> WalletMetadata -> MockOp ()
+mCreateWallet wid wal meta m@(M cp metas txs pk)
+    | wid `Map.member` cp = (Left (WalletAlreadyExists wid), m)
+    | otherwise =
+        ( Right ()
+        , M
+            { checkpoints = Map.insert wid wal cp
+            , metas = Map.insert wid meta metas
+            , txs = txs
+            , privateKey = pk
+            }
+        )
+
+mRemoveWallet :: MWid -> MockOp ()
+mRemoveWallet wid m@(M cp metas txs pk)
+    | wid `Map.member` cp =
+        ( Right ()
+        , M
+            { checkpoints = Map.delete wid cp
+            , metas = Map.delete wid metas
+            , txs = Map.delete wid txs
+            , privateKey = Map.delete wid pk
+            }
+        )
+    | otherwise = (Left (NoSuchWallet wid), m)
+
+mListWallets :: MockOp [MWid]
+mListWallets m@(M cp _ _ _) = (Right (L.sortOn unMockWid $ Map.keys cp), m)
+
+mPutCheckpoint :: MWid -> MWallet -> MockOp ()
+mPutCheckpoint wid wal m@(M cp metas txs pk)
+    | wid `Map.member` cp = (Right (), M (Map.insert wid wal cp) metas txs pk)
+    | otherwise = (Left (NoSuchWallet wid), m)
+
+mReadCheckpoint :: MWid -> MockOp (Maybe MWallet)
+mReadCheckpoint wid m@(M cp _ _ _) = (Right (Map.lookup wid cp), m)
+
+mPutWalletMeta :: MWid -> WalletMetadata -> MockOp ()
+mPutWalletMeta wid meta m@(M cp metas txs pk)
+    | wid `Map.member` cp = (Right (), M cp (Map.insert wid meta metas) txs pk)
+    | otherwise = (Left (NoSuchWallet wid), m)
+
+mReadWalletMeta :: MWid -> MockOp (Maybe WalletMetadata)
+mReadWalletMeta wid m@(M _ meta _ _) = (Right (Map.lookup wid meta), m)
+
+mPutTxHistory :: MWid -> TxHistory -> MockOp ()
+mPutTxHistory wid txs' m@(M cp metas txs pk)
+    | wid `Map.member` cp = (Right (), M cp metas (Map.alter appendTxs wid txs) pk)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+    where appendTxs = Just . (<> txs') . fromMaybe mempty
```

Comment:
Map union is _left-biased_, which means that in case of conflicting keys, the entries from the left map will prevail. However, we've given another semantic to our database-layer: new transaction should overwrite existing ones (debatable, see below). I wondered however why the tests failed to catch that as I remember making sure to generate transaction ids with a rather small entropy. 

Yet, the `Arbitrary` instance for `Hash "Tx"` has been changed to generate 32-byte hashes across the whole range, which makes 16^32 (2^128) different hashes, and therefore, we probably never generate twice the same tx id.

Changing this to something that generates more conflict indeed yield the following error:

```hs
       uncaught exception: SqliteException
       SQLite3 returned ErrorConstraint while attempting to perform step: UNIQUE constraint failed: tx_out.tx_id, tx_out.index
       (after 89 tests and 189 shrinks)
```

Now the question about whether the semantic we initially gave is correct is a valid question. The rationale was that, during rollbacks, we will have to possibly replace metadata of some existing transactions. One approach could be to overwrite transactions with some others in order to avoid having to perform a delete and then an insert. However, there's some values in making sure we correctly clean-up the database, throwing an exception if we try to insert twice a transaction -- this would clearly alert us that something's broken in our cleanup process.

What do you think @rvl ? Also, if we keep the generator as it is, it may be worth adding a precondition to verify that we don't already know of a transaction with a newly generated id. Conflicts are very unlikely over 2^128, but not impossible provided we generate quite many transactions on quite many runs...